### DISCUSSION TITLE
Extracting multiple data with parse

### DISCUSSION THREAD

**Discussion by SebastianMpl** [Oldes/Rebol3#93] [Q&A] [answered]
Hello,
Today I'm writing because of issue I have with parse. I would like to extract all the numbers from a string.
In a book Learn Red Fundamentals i found a code that I hoped will solve this and modified it for multiple data:
```
nums: "34 56 83 19"
digits: charset [#"0"-#"9"]
ints: copy []
rules: [some [copy int some digits | skip](append ints int)] 
```
The result of parse nums rules is 
>> ints
== ["19"]
So only the last number is saved.
I also tried with 
`rules: [any [some [copy int some digits | skip]](append ints int)]`

But the result is the same. What am I missing? Could you please help?
Thank you

**Comment by Oldes** [comment, marked as answer]
```rebol
nums: "34 56 83 19"
digits: system/catalog/bitsets/numeric
parse nums [collect any [keep some digits | skip]]
;== ["34" "56" "83" "19"]
```
or in your case:
```rebol
ints: copy []
rules: [some [copy int some digits (append ints int)| skip]]
parse nums rules
ints
;== ["34" "56" "83" "19"]
```
Your error is, that you are appending just the last `int`, having `(append ints int)` outside the `some` block.

**Reply by SebastianMpl** [reply]
Thank you very much !
And for other code also. Many things to study.

**Comment by Oldes** [comment]
Also if your input is just numbers with spaces as a delimiter, you can use:
```rebol
split "34 56 83 19" SP
;== ["34" "56" "83" "19"]
```

**Comment by Oldes** [comment]
If you need integers directly from the `parse`, you can use:
```rebol
collect [parse nums [some [copy int some digits (keep to integer! int) | skip]]]
;== [34 56 83 19]
```
(If you have it in a function, don't forget to define `int` as a `/local`)

**Comment by Oldes** [comment]
Another better way:
```rebol
parse nums [collect any [copy int some digits keep (to integer! int) | skip]]
== [34 56 83 19]
```
As `collect` as a `parse` command is native!
```
Running 2 code blocks 1000 times.
--------------------------------------------------------------------------------------------------------------------------------------------------
Time               | Evals  | S.made | S.expa | Memory      | Code
1.0x (5μs)         | 58     | 7      | 2      | 320         | [parse nums [collect any [copy int some digits keep (to integer! int) | skip]]]
2.81x (15μs)       | 117    | 20     | 0      | 2464        | [collect [parse nums [some [copy int some digits (keep to integer! int) | skip]]]]
--------------------------------------------------------------------------------------------------------------------------------------------------
